reframe/utility/typecheck.py

```diff
@@ -93,24 +93,35 @@
 import re
 
 
-class _TypeFactory(abc.ABCMeta):
-    def register_subtypes(cls):
-        for t in cls._subtypes:
-            cls.register(t)
+class Type(abc.ABCMeta):
+    def __call__(cls, *args, **kwargs):
+        if (len(args) == 1 and
+            not kwargs and isinstance(args[0], str) and
+            hasattr(cls, '__rfm_cast__')):
+            return cls.__rfm_cast__(*args)
+        else:
+            obj = cls.__new__(cls, *args, **kwargs)
+            obj.__init__(*args, **kwargs)
+            return obj
 
 
-# Metaclasses that implement the isinstance logic for the different aggregate
-# types
+# Metaclasses that implement the isinstance logic for the different builtin
+# container types
 
-class _ContainerType(_TypeFactory):
+class _ContainerType(Type):
+    def register_container_type(cls):
```

Comment:
Yes, this used to be a bit more complex before when instead of `_type` I had `_subtypes` which was a tuple.

Reply:
I haven't put a comment eventually, because I don't think it's needed. I've now put a check that `_type` is defined and along with the name of the function, it should be self explaining now. I've also removed some redundant calls to this method elsewhere.